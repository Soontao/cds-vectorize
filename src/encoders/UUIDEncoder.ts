import { Encoder } from "./Encoder.js";
import { isNull, nullVector } from "./util.js";

export class UUIDEncoder implements Encoder<string> {
  /**
   *
   * @param hexString
   * @returns
   */
  #hexToUint32Array(hexString: string) {
    const buffer = Buffer.from(hexString, "hex");

    const numElements = Math.ceil(buffer.length / Uint32Array.BYTES_PER_ELEMENT);

    return Array.from(new Uint32Array(buffer.buffer, buffer.byteOffset, numElements));
  }

  #Uint32ArrayToHex(arr: Array<number>) {
    return Buffer.from(new Uint32Array(arr).buffer).toString("hex");
  }

  features(name: string): string[] {
    return [`${name}_1`, `${name}_2`, `${name}_3`, `${name}_4`];
  }

  encode(value: string): number[] {
    if (isNull(value)) {
      return nullVector(this.length);
    }
    value = value.replaceAll("-", "");
    return this.#hexToUint32Array(value);
  }

  decode(vec: number[]): string {
    const hex = this.#Uint32ArrayToHex(vec);
    return [hex.slice(0, 8), hex.slice(8, 12), hex.slice(12, 16), hex.slice(16, 20), hex.slice(20)].join("-");
  }

  get length(): number {
    return 4;
  }
}

export default UUIDEncoder;
