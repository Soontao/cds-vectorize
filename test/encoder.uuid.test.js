import { randomUUID } from "crypto";
import { UUIDEncoder } from "../src/encoders/UUIDEncoder.js";

describe('UUIDEncoder Test Suite', () => {

  it('should support UUID encoder', () => {
    const e = new UUIDEncoder()
    const u = randomUUID()
    const vec = e.encode(u)

    expect(vec).toHaveLength(4)
    expect(e.decode(vec)).toEqual(u)

  });

  it('should support UUID encoder with snapshot test', () => {
    const e = new UUIDEncoder()
    const v = e.encode('d59a5c43-2e7a-445e-ac3c-b4ed862da041')
    expect(v).toMatchSnapshot()
    expect(e.decode(v)).toMatchSnapshot()
  });

  // AI
  it("encodes and decodes a valid UUID", () => {
    const encoder = new UUIDEncoder();
    const uuid = "550e8400-e29b-41d4-a716-446655440000";
    const encodedValue = encoder.encode(uuid);
    const decodedValue = encoder.decode(encodedValue);
    expect(decodedValue).toBe(uuid);
  });

  it("encodes and decodes an empty UUID", () => {
    const encoder = new UUIDEncoder();
    const emptyUUID = "00000000-0000-0000-0000-000000000000";
    const encodedEmptyUUID = encoder.encode(emptyUUID);
    const decodedEmptyUUID = encoder.decode(encodedEmptyUUID);
    expect(decodedEmptyUUID).toBe(emptyUUID);
  });

  it("should encode a UUID string to a Uint32Array", () => {
    const encoder = new UUIDEncoder();
    const uuid = "550e8400-e29b-41d4-a716-446655440000";
    const expectedArray = [
      8654421,
      3561069538,
      1715738279,
      17493,
    ];

    const result = encoder.encode(uuid);

    expect(result).toEqual(expectedArray);
  });

  it("should decode a Uint32Array to a UUID string", () => {
    const encoder = new UUIDEncoder();
    const uuid = "58319155-6b94-0000-3442-000052410000";
    const inputArray = [
      1435578712, 37995, 16948, 16722,
    ];

    const result = encoder.decode(inputArray);

    expect(result).toEqual(uuid);
  });

  it("should return an array with the given name in features", () => {
    const encoder = new UUIDEncoder();
    const name = "myUUID";
    const result = encoder.features(name);

    expect(result).toHaveLength(4);
    expect(result).toMatchSnapshot();
  });

  it("should return the correct length of the encoded array", () => {
    const encoder = new UUIDEncoder();
    const result = encoder.length;

    expect(result).toEqual(4);
  });

});
