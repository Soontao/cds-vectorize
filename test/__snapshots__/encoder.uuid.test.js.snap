// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UUIDEncoder Test Suite should return an array with the given name in features 1`] = `
[
  "myUUID_1",
  "myUUID_2",
  "myUUID_3",
  "myUUID_4",
]
`;

exports[`UUIDEncoder Test Suite should support UUID encoder with snapshot test 1`] = `
[
  1130142421,
  1581546030,
  3988012204,
  1101016454,
]
`;

exports[`UUIDEncoder Test Suite should support UUID encoder with snapshot test 2`] = `"d59a5c43-2e7a-445e-ac3c-b4ed862da041"`;
